crates/pathfinder/src/retry.rs

```diff
@@ -0,0 +1,351 @@
+where
+    Fut: Future<Output = Result<T, E>>,
+    FutureFactory: FnMut() -> Fut,
+{
+    /// Create an exponential [`Retry`] utility for a future which is created by `future_factory`
+    /// with initial backoff of `base_secs` seconds.
+    ///
+    /// `Nth` backoff is equal to `base_secs ^ N` seconds.
+    pub fn exponential(future_factory: FutureFactory, base_secs: NonZeroU64) -> Self {
+        Self {
+            future_factory,
+            strategy: Strategy {
+                initial_backoff_secs: base_secs,
+                factor: NonZeroU64::new(1).unwrap(),
+                max_delay_secs: None,
+                max_num_retries: None,
+            },
+        }
+    }
+
+    /// Multiply backoff by this factor.
+    ///
+    /// `Nth` backoff is then equal to `base_secs ^ N * factor` seconds.
+    pub fn factor(mut self, factor: NonZeroU64) -> Self {
+        self.strategy.factor = factor;
+        self
+    }
+
+    /// Saturate backoff at `max_delay_secs` seconds.
+    pub fn max_delay_secs(mut self, max_delay_secs: NonZeroU64) -> Self {
+        self.strategy.max_delay_secs = Some(max_delay_secs);
+        self
+    }
+
+    /// Limit the number of retries to `max_num_retries`.
+    pub fn max_num_retries(mut self, max_num_retries: NonZeroUsize) -> Self {
+        self.strategy.max_num_retries = Some(max_num_retries);
+        self
+    }
+
+    /// Retry the future on any `Err()` until an `Ok()` value is returned by the future.
+    pub async fn on_any_err(self) -> Result<T, E> {
+        TokioRetry::spawn(MaybeLimited::from(self.strategy), self.future_factory).await
+    }
+
+    /// Retry the future on every error that meets `retry_condition` until the future returns:
+    /// - an `Ok()` value
+    /// - an `Err()` value that does not meet the `retry_condition`.
+    pub async fn when<RetryCondition>(self, retry_condition: RetryCondition) -> Result<T, E>
+    where
+        RetryCondition: FnMut(&E) -> bool,
+    {
+        TokioRetryIf::spawn(
+            MaybeLimited::from(self.strategy),
+            self.future_factory,
+            retry_condition,
+        )
+        .await
+    }
+}
+
+struct Strategy {
+    initial_backoff_secs: NonZeroU64,
+    factor: NonZeroU64,
+    max_delay_secs: Option<NonZeroU64>,
+    max_num_retries: Option<NonZeroUsize>,
+}
+
+enum MaybeLimited {
+    Limited(std::iter::Take<ExponentialBackoff>),
+    Unlimited(ExponentialBackoff),
+}
+
+impl std::iter::Iterator for MaybeLimited {
+    type Item = std::time::Duration;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            MaybeLimited::Limited(x) => x.next(),
+            MaybeLimited::Unlimited(x) => x.next(),
+        }
+    }
+}
+
+impl From<Strategy> for MaybeLimited {
+    fn from(s: Strategy) -> Self {
+        // We use milliseconds in tests
+        #[cfg(test)]
+        const FACTOR: u64 = 1;
+
+        // We use seconds in production
+        #[cfg(not(test))]
+        const FACTOR: u64 = 1000;
```

Comment:
Another option is to use [tokio's advance](https://docs.rs/tokio/latest/tokio/time/fn.advance.html) function which lets you fake time skips in tests (I'm assuming this is done to have the tests run in some reasonable time).

Reply:
This is awesome :+1: lemme try it.

Reply:
I tried it but unfortunately it does not allow to assert for expected delays.